### DISCUSSION TITLE
Craete account?

### DISCUSSION THREAD

**Discussion by numairawan** [qgustavor/mega#127] [Ideas]
How can i use this lib to create an account?

**Comment by qgustavor** [comment]
That's not supported. I'll move this to a discussion and label it as an idea.

**Comment by qgustavor** [comment]
[It was planned in 2016](https://github.com/qgustavor/mega/issues/8) but I gave up from implementing it. One of the reasons I never implemented it is that, at the moment, the library does not have functions to generate RSA keys which are used to authenticate with MEGA, just functions to load existent keys and authenticate. The key format used to store RSA keys is non-standard format, [I tried working with that](https://github.com/qgustavor/mega/issues/3), without success. Probably it should be easier to implement nowadays since Node.js supports WebCrypto and WebCrypto supports RSA functions.

If someone wants to work on this idea feel free to send a pull request.

**Comment by numairawan** [comment]
[Here ](https://stackoverflow.com/questions/38155000/generate-rsa-keys-using-javascript?answertab=modifieddesc#tab-top) in Web Crypto we can actually generate the RSA keys. I am not sure how to implement as i didn't study the code yet but if you explain me bit where to start then i can work on it. 

And for a nodejs this [might ](https://nkhilv.medium.com/generating-a-private-rsa-key-for-tests-node-js-c3fc23b3ef1c) help.

**Reply by qgustavor** [reply]
Creating the keys with WebCrypto is kinda easy. Here's what I think that's needed:

1. Check some other client code to check how accounts are created
    - The official client sometimes is too messy because whoever wrote that code thought using jQuery was a good idea, so I often prefer checking megatools source-code
1. Find a way to convert RSA keys from WebCrypto's formats to MEGA's format and vice-versa
    - [This issue](https://github.com/qgustavor/mega/issues/3) may help on that
1. Find about MEGA API registration endpoints:
    - One that will send an message to an email address with some info to validate that address and another to confirm that
    - One of those two should also accept your public key, the private key encrypted with the hashed user password, the salt used to hash the password and [a stringhash](https://github.com/qgustavor/mega/blob/6353191717694f6857164ae6548dfc92254763eb/lib/crypto/aes.mjs#L62-L85) to enable authentication with the server.
1. Design the library side of this:
    - An static method on `Storage` like `Storage.registerAccount`? A new top-level export? I prefer a new top-level export to make it tree shakeable.
    - How should email confirmation work? That above function return something that can be called to continue the registration process or use a second method?
1. Write tests for it (if possible)
    - It should require implementing those methods in the API mocking server and implementing support for V2 accounts on it
